staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -207,7 +207,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 		return false, errors.New("cloud provider does not support instances")
 	}
 
-	shutdown, err := instances.InstanceShutdownByProviderID(ctx, node.Spec.ProviderID)
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))
```

Comment:
Isn't [getProviderID](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L434) the method you should be using here?

Reply:
Maybe it's a duplicated implementation. It is just copied from `ensureNodeExistsByProviderID`  which is in the same file

https://github.com/kubernetes/kubernetes/blob/5ab5300a3aae230162b676eda4520f81c0923115/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L237C6-L263

Reply:
But the main difference is that we are ignoring the error `InstanceNotFound` which we should as we are in the function `shutdownInCloudProvider`. This should not happen if it is already deleted.

Reply:
Ah, I see. However, looks like the result is going to be different between call `instances.InstanceID(ctx, types.NodeName(node.Name))` and `cnc.getProviderID(ctx, copyNode)`, right?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/cloud.go#L101-L115

Reply:
Ah, looks like you discuss in another comment. Seems like a bug that it's using `instances.InstanceID(ctx, types.NodeName(node.Name))` anywhere to get `providerID`?

Reply:
I think we should use both. Get the provider ID via `getProviderID` and ignore the error if no instance is found.
Will update the PR tomorrow

Reply:
Makes sense. If we're getting the provider ID, we should get it properly.

Reply:
One problem: the function `GetInstanceProviderID` is returning a different error to the const `InstanceNotFound` if the instance is not found

`InstanceNotFound`
https://github.com/kubernetes/kubernetes/blob/2f270bd99609561be53893d3c7fc77d8efe7ac4e/staging/src/k8s.io/cloud-provider/cloud.go#L256

general error of `GetInstanceProviderID`
https://github.com/kubernetes/kubernetes/blob/2f270bd99609561be53893d3c7fc77d8efe7ac4e/staging/src/k8s.io/cloud-provider/cloud.go#L112

But I will change the code that a real provider ID is used

Reply:
I created a method `getProviderID`, WDYT?